src/main/scala/com/fulcrumgenomics/commons/async/AsyncRunnable.scala

```diff
@@ -0,0 +1,168 @@
+  def done: Boolean = this.doneLatch.getCount == 0
+
+  /** Returns a throwable if an exception occurred in the [[run()]] method, None otherwise. */
+  def throwable: Option[Throwable] = Option(this._throwable.get())
+
+  /** Waits for the [[run()]] method to start. */
+  def awaitStart(): Unit = this.startedLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has started.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def startAwaitable()(implicit ec: ExecutionContext) : Awaitable[Unit] = Future { this.awaitStart() }
+
+  /** Waits for the [[run()]] method to complete. */
+  def awaitDone(): Unit = this.doneLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has completed.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def doneAwaitable()(implicit ec: ExecutionContext) : Awaitable[Option[Throwable]] = Future { this.awaitDone(); Option(this._throwable.get()) }
+
+  /** Creates a new thread wrapping this runnable; the thread is not started.
+    *
+    * @param name optionally the name of the thread, otherwise a name is created based on the class name and the number
+    *             of threads already created.
+    */
+  final def thread(name: Option[String] = None): Thread = {
+    require(!this.started, "Already started.")
+    val threadName = name.getOrElse { AsyncRunnable.nextName(this) }
+    new Thread(this, threadName)
+  }
+
+  /** Starts this [[Runnable]] in a daemon thread.
+    *
+    * @param name optionally the name of the thread, otherwise a name is created based on the class name and the number
+    *             of threads already created.
+    */
+  final def start(name: Option[String] = None): this.type = {
+    require(!this.started, "Already started.")
+    val t = thread(name)
+    t.setDaemon(true)
+    t.start()
+    this.startedLatch.countDown()
+    this
+  }
+
+  /** The method that does the asynchronous work. */
+  protected def execute(): Unit
+
+  /** The method to execute if an exception occurs in the asynchronous thread.  This should not block. */
+  protected def uponException(): Unit = Unit
+
+  /** The method to execute upon successfully execution of the run method or an exception occurs.  This should not block. */
+  protected def uponFinally(): Unit = Unit
+
+  /** Checks to see if an exception has been raised by an asynchronous thread and if so rethrows it as an [[Error]]
+    * or [[RuntimeException]] as appropriate.  Use this method before code that assumes the threads have not encountered
+    * an exception.
+    */
+  protected final def checkAndRaise(): Unit = {
+    _throwable.getAndSet(null) match {
+      case null                 => Unit
+      case e: Error             => throw e
+      case ex: RuntimeException => throw ex
+      case t                    => throw new RuntimeException(t)
```

Comment:
I'm curious why this is necessary?  All exceptions are unchecked in scala...

Reply:
You're right, so it can be:
```
    _throwable.getAndSet(null) match {
      case null           => Unit
      case thr: Throwable => throw thr
    }
```